src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala

```diff
@@ -76,47 +77,55 @@ object ErgoState extends ScorexLogging {
-      case (tx :: tail, r: Valid[Long]) =>
-        val boxesToSpendTry: Try[List[ErgoBox]] = tx.inputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
-
-        lazy val dataBoxesTry: Try[List[ErgoBox]] = tx.dataInputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
-
-        lazy val boxes: Try[(List[ErgoBox], List[ErgoBox])] = dataBoxesTry.flatMap(db => boxesToSpendTry.map(bs => (db, bs)))
-
-        val vs = tx.validateStateless()
-          .validateNoFailure(txBoxesToSpend, boxesToSpendTry)
-          .validateNoFailure(txDataBoxes, dataBoxesTry)
-          .payload[Long](r.value)
-          .validateTry(boxes, e => ModifierValidator.fatal("Missed data boxes", e)) { case (_, (dataBoxes, toSpend)) =>
-            tx.validateStateful(toSpend.toIndexedSeq, dataBoxes.toIndexedSeq, currentStateContext, r.value)(verifier).result
-          }
-
-        execTx(tail, vs)
-      case _ =>
-        accCostTry
+    def collectBoxesById(
+                 remainingBoxIds: Iterator[ErgoBox.BoxId],
+                 resultingBoxes: Try[VectorBuilder[ErgoBox]] = Success(new VectorBuilder[ErgoBox])
+               ): Try[IndexedSeq[ErgoBox]] = {
+      if (!remainingBoxIds.hasNext) {
+        resultingBoxes.map(_.result())
+      } else {
+        checkBoxExistence(remainingBoxIds.next()) match {
+          case Success(box) =>
+            collectBoxesById(remainingBoxIds, resultingBoxes.map(_ += box))
+          case Failure(ex) =>
+            Failure(ex)
+        }
+      }
     }
 
     // Skip v1 block transactions validation if corresponding setting is on
     if (currentStateContext.blockVersion == 1 &&
           currentStateContext.ergoSettings.nodeSettings.skipV1TransactionsValidation) {
       Valid(0L)
     } else {
-      execTx(transactions.toList, Valid[Long](0L))
+      import spire.syntax.all.cfor
+      var costResult: ValidationResult[Long] = Valid[Long](0L)
+      cfor(0)(_ < transactions.length && costResult.isValid, _ + 1) { i =>
+        val validCostResult = costResult.asInstanceOf[Valid[Long]]
+        val tx = transactions(i)
+        val boxesToSpendTry: Try[IndexedSeq[ErgoBox]] = collectBoxesById(tx.inputs.iterator.map(_.boxId))
+        lazy val dataBoxesTry: Try[IndexedSeq[ErgoBox]] = collectBoxesById(tx.dataInputs.iterator.map(_.boxId))
+        lazy val boxes: Try[(IndexedSeq[ErgoBox], IndexedSeq[ErgoBox])] = dataBoxesTry.flatMap(db => boxesToSpendTry.map(bs => (db, bs)))
+        costResult = tx.validateStateless()
+          .validateNoFailure(txBoxesToSpend, boxesToSpendTry)
+          .validateNoFailure(txDataBoxes, dataBoxesTry)
+          .payload[Long](validCostResult.value)
+          .validateTry(boxes, e => ModifierValidator.fatal("Missed data boxes", e)) { case (_, (dataBoxes, toSpend)) =>
+            tx.validateStateful(toSpend, dataBoxes, currentStateContext, validCostResult.value)(verifier).result
```

Comment:
The currentStateContext is not update after each tx, is it ok?

Reply:
it is, just a few lines above `costResult = tx.validateStateless()`